src/Lean/Class.lean

```diff
@@ -84,7 +84,10 @@ def addClass (env : Environment) (clsName : Name) : Except String Environment :=
     | some decl@(ConstantInfo.inductInfo _) => do
       let b ← checkOutParam 1 #[] decl.type
       Except.ok (classExtension.addEntry env { name := clsName, hasOutParam := b })
-    | some _    => Except.error ("invalid 'class', declaration '" ++ toString clsName ++ "' must be inductive datatype or structure")
+    | some decl@(ConstantInfo.axiomInfo _) => do
+      let b ← checkOutParam 1 #[] decl.type
+      Except.ok (classExtension.addEntry env { name := clsName, hasOutParam := b })
```

Comment:
What's the sleekest way to avoid the code duplication here?

Reply:
How about
```lean
def addClass (env : Environment) (clsName : Name) : Except String Environment := do
  if isClass env clsName then
    throw s!"class has already been declared '{clsName}'"
  let some decl := env.find? clsName
    | throw s!"unknown declaration '{clsName}'"
  unless decl matches .inductInfo .. || decl matches .axiomInfo .. do
    throw s!"invalid 'class', declaration '{clsName}' must be inductive datatype, structure, or constant"
  let b ← checkOutParam 1 #[] decl.type
  return classExtension.addEntry env { name := clsName, hasOutParam := b }
```